### PR TITLE
Akemi/auth 3958 privateapps tf

### PR DESCRIPTION
**Pull request by AkemiDavisson**
Adds private app type and related fields

Requires https://github.com/cloudflare/cloudflare-go/pull/762 before merge

### CODE REVIEW

File: cloudflare/resource_cloudflare_access_application_test.go [hunk 1]
```diff
@@ -424,6 +424,33 @@ resource "cloudflare_access_application" "%[1]s" {
 `, rnd, zoneID, domain, accountID)
 }
 
+func testAccCloudflareAccessApplicationConfigWithGatewayRules(rnd, zoneID, domain string, accountID string) string {
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_access_application_test.go:427]: @AkemiDavisson looks like this is unused and we don't have any test coverage for these new fields. do you mind updating this?

Comment (AkemiDavisson) [reply cloudflare/resource_cloudflare_access_application_test.go:456]: Updated! 👍

File: cloudflare/resource_cloudflare_access_application.go [hunk 2]
```diff
@@ -43,6 +43,16 @@ func resourceCloudflareAccessApplicationCreate(d *schema.ResourceData, meta inte
 		LogoURL:                 d.Get("logo_url").(string),
 		SkipInterstitial:        d.Get("skip_interstitial").(bool),
 		AppLauncherVisible:      d.Get("app_launcher_visible").(bool),
+		PrivateAddress:          d.Get("private_address").(string),
+	}
+
+	gatewayRules := d.Get("gateway_rules").([]cloudflare.AccessApplicationGatewayRule)
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_access_application.go:49]: looks like this is causing issues with the acceptance tests as we are casting to the wrong type. i think you want `[]interface{}` here.

```
panic: interface conversion: interface {} is []interface {}, not []cloudflare.AccessApplicationGatewayRule

goroutine 7978 [running]:
github.com/cloudflare/terraform-provider-cloudflare/cloudflare.resourceCloudflareAccessApplicationUpdate(0xc000311200, {0x1a8b600, 0xc0001a2480})
	/Users/jacob/go/src/github.com/cloudflare/terraform-provider-cloudflare/cloudflare/resource_cloudflare_access_application.go:178 +0xcc6
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0x197ec40, {0x1c0cbb8, 0xc0004a0500}, 0x24, {0x1a8b600, 0xc0001a2480})
	/Users/jacob/.asdf/installs/golang/1.17.6/packages/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.10.1/helper/schema/resource.go:363 +0x178
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc0003a2540, {0x1c0cbb8, 0xc0004a0500}, 0xc000334680, 0xc000311080, {0x1a8b600, 0xc0001a2480})
	/Users/jacob/.asdf/installs/golang/1.17.6/packages/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.10.1/helper/schema/resource.go:475 +0x6ba
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc0007ec030, {0x1c0cbb8, 0xc0004a0500}, 0xc000534140)
	/Users/jacob/.asdf/installs/golang/1.17.6/packages/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.10.1/helper/schema/grpc_provider.go:977 +0xd8a
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc00068c580, {0x1c0cc60, 0xc0005661e0}, 0xc000a16000)
	/Users/jacob/.asdf/installs/golang/1.17.6/packages/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.5.0/tfprotov5/tf5server/server.go:603 +0x30e
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x1a5c820, 0xc00068c580}, {0x1c0cc60, 0xc0005661e0}, 0xc0009563c0, 0x0)
	/Users/jacob/.asdf/installs/golang/1.17.6/packages/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.5.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:380 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00069a1c0, {0x1c1eb30, 0xc000202780}, 0xc000a10000, 0xc00076c780, 0x2235fc0, 0x0)
	/Users/jacob/.asdf/installs/golang/1.17.6/packages/pkg/mod/google.golang.org/grpc@v1.40.0/server.go:1297 +0xccf
google.golang.org/grpc.(*Server).handleStream(0xc00069a1c0, {0x1c1eb30, 0xc000202780}, 0xc000a10000, 0x0)
	/Users/jacob/.asdf/installs/golang/1.17.6/packages/pkg/mod/google.golang.org/grpc@v1.40.0/server.go:1626 +0xa2a
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	/Users/jacob/.asdf/installs/golang/1.17.6/packages/pkg/mod/google.golang.org/grpc@v1.40.0/server.go:941 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/Users/jacob/.asdf/installs/golang/1.17.6/packages/pkg/mod/google.golang.org/grpc@v1.40.0/server.go:939 +0x294
FAIL	github.com/cloudflare/terraform-provider-cloudflare/cloudflare	165.245s
?   	github.com/cloudflare/terraform-provider-cloudflare/tools/cmd/changelog-check	[no test files]
?   	github.com/cloudflare/terraform-provider-cloudflare/version	[no test files]
FAIL
make: *** [testacc] Error 1
```

File: cloudflare/resource_cloudflare_access_application.go [hunk 3]
```diff
@@ -148,6 +171,16 @@ func resourceCloudflareAccessApplicationUpdate(d *schema.ResourceData, meta inte
...
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_access_application.go:177]: same as above

File: cloudflare/schema_cloudflare_access_application.go [hunk 4]
```diff
@@ -153,6 +153,22 @@ func resourceCloudflareAccessApplicationSchema() map[string]*schema.Schema {
 			Optional: true,
 			Default:  true,
 		},
+		"gateway_rules": {
+			Type:     schema.TypeList,
```
Comment (jacobbednarz) [inline cloudflare/schema_cloudflare_access_application.go:157]: does ordering matter? if not, you may want `TypeSet` instead.

**Review by jacobbednarz — COMMENTED** [review]
sorry on the delay; this totally slipped off my radar. i've added a couple of notes inline after trying to run the tests and hitting some type issues.

let me know if you need a hand getting this over the line!

**Pull request closed without merging**